### PR TITLE
FEATURE - Add range port

### PR DESCRIPTION
**Pull request by LOLOLEKIK**
# Description

Code modified to expose port ranges rather than just ports.

In fact, in environments where it's not possible to have a bridged network, it's a pain to expose several ports at once.

The port exposure feature has been changed to allow commands of this kind: 

exegol start dev -p 9000-9010:9000-9010 -p 80 -p 802-805:udp

# Point of attention

The feature has been extensively tested, but I have two points to make.

1. Creation timeout

First of all, if a port range is too large, it is possible to be timeout by exegol.

More precisely, if the machine takes longer than 60 seconds to create, exegol stops. I didn't touch this behavior to avoid blocking the user. In this case, the long creation time is not linked to exegol but to docker.

It's starting to be difficult to create a container with more than 2500 published ports.

Once again, this is specific to docker and exegol would behave in the same way as if you used the --ports option 2500 times. 

ref : 
https://forums.docker.com/t/i-have-a-docker-container-that-needs-to-expose-10-000-ports/96048/3
https://github.com/moby/moby/issues/14288

2. Display port

Similar effect on port display. Ports are displayed in sequence and not by range. For standard use, there's no visual imapact, but for exposing hundreds of ports, this has an impact on the visibility of the creation summary.

I did some local tests to manage the display a little better. 

But I find the current interface more reassuring (ports are clearly visible).

To sum up, the modification also meant a significant increase in the complexity of the code and a less clear interface.

### CODE REVIEW

File: exegol/model/ContainerConfig.py [hunk 1]
```diff
@@ -1234,26 +1234,41 @@ def addUserDevice(self, user_device_config: str):
...
         If host_ipv4 is not set, default to 0.0.0.0
-        If container_port is not set, default is the same as host port
+        If container_port is not set, the same port(s) as host port(s) will be used
         If protocol is not set, default is 'tcp'"""
-        match = re.search(r"^((\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}):)?(\d+):?(\d+)?:?(udp|tcp|sctp)?$", user_test_port)
+        # Regex to capture port ranges and protocols correctly
+        match = re.search(r"^((\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}):)?(\d+)(-(\d+))?:?(\d+)?(-(\d+))?:?(udp|tcp|sctp)?$", user_test_port)
         if match is None:
-            logger.critical(f"Incorrect port syntax ({user_test_port}). Please use this format: [green][<host_ipv4>:]<host_port>[:<container_port>][:<protocol>][/green]")
+            logger.critical(f"Incorrect port syntax ({user_test_port}). Please use the correct format.")
```
Comment (Dramelac) [inline exegol/model/ContainerConfig.py:1245]: Please keep the updated format reminder here

File: exegol/model/ContainerConfig.py [hunk 2]
```diff
@@ -1234,26 +1234,41 @@ def addUserDevice(self, user_device_config: str):
...
         host_ip = "0.0.0.0" if match.group(2) is None else match.group(2)
-        protocol = "tcp" if match.group(5) is None else match.group(5)
-        try:
-            host_port = int(match.group(3))
-            container_port = host_port if match.group(4) is None else int(match.group(4))
-            if host_port > 65535 or container_port > 65535:
-                raise ValueError
-        except ValueError:
-            logger.critical(f"The syntax for opening prot in NAT is incorrect. The ports must be numbers between 0 and 65535. ({match.group(3)}:{match.group(4)})")
+        start_host_port = int(match.group(3))
```
Comment (Dramelac) [inline exegol/model/ContainerConfig.py:1249]: Those `int(<user_input>)` can raise `ValueError` exception if the user doesn't input number in the field, you must keep the try/except block (it wasn't here just to check port over 65535).

File: exegol/model/ContainerConfig.py [hunk 3]
```diff
@@ -1234,26 +1234,41 @@ def addUserDevice(self, user_device_config: str):
...
+        if end_host_port > 65535 or end_container_port > 65535:
+            logger.critical(f"The syntax for opening prot in NAT is incorrect. The ports must be numbers between 0 and 65535. ({end_host_port}:{end_container_port})")
             return
-        self.addPort(host_port, container_port, protocol=protocol, host_ip=host_ip)
+
+        try:
+            for host_port, container_port in zip(range(start_host_port, end_host_port + 1), range(start_container_port, end_container_port + 1)):
+                self.addPort(host_port, container_port, protocol=protocol, host_ip=host_ip)
+        except ValueError as e:
+            logger.critical(f"The syntax for opening prot in NAT is incorrect. The format must be [<host_ipv4>:]<host_port>[-<end_host_port>][:<container_port>[-<end_container_port>]][:<protocol>] ({match.group(3)}:{match.group(4)})")
```
Comment (Dramelac) [inline exegol/model/ContainerConfig.py:1269]: Why put the try/except here ? It should by on port parsing.
(Plus, I just saw a typo here `prot`)

**Review by Dramelac — CHANGES_REQUESTED** [review]
Hello, 
Thank you for your PR ! That would be a nice addition.

Here is a first review focused on your code changes.

Can you add a screenshot of the recap table when there is a lot of port shared ? Maybe we can add a post-process logic to the function getTextPorts() of ContainerConfig to merge those range ?

File: exegol/model/ContainerConfig.py (hunk 1 above)
Comment (LOLOLEKIK) [reply exegol/model/ContainerConfig.py]: My bad. It's fix

File: exegol/model/ContainerConfig.py (hunk 2 above)
Comment (LOLOLEKIK) [reply exegol/model/ContainerConfig.py]: Okay, I hadn't thought of it that way. I put back the try/except

File: exegol/model/ContainerConfig.py (hunk 3 above)
Comment (LOLOLEKIK) [reply exegol/model/ContainerConfig.py]: I put it there out of habit, but it shouldn't be necessary :) (typo fix)

**Comment by LOLOLEKIK** [comment]
Here's a screenshot of the acutel result. As I said, I found it less clear to have the port range in the summary. For standard use, the output remains readable as you can see. If hundreds of ports are open, the terminal will just scroll.

image

File: exegol/model/ContainerConfig.py (hunk 1 above)
Comment (Dramelac) [reply exegol/model/ContainerConfig.py]: Can you also keep the `[green][/green]` text coloring format ?

Comment (LOLOLEKIK) [reply exegol/model/ContainerConfig.py]: Done, tell me if it's not ok.

**Comment by ShutdownRepo** [comment]
I love the addition, thank you @LOLOLEKIK :muscle: 
@Dramelac will handle the review of this Pull Request, but please don't forget to run some fuzzy testing to make sure that messed up entries still get handled. For instance :point_down: 
```
# left array length < right array length
-p 1000-1010:2000-2999  

 # left int vs. right array
-p 3000:3000-3005 

# port out of bound
-p 99999 

# incomplete array 
-p 4000-  

# would this work?
-p  5000,5001,5002,5003-5005,5010-5015  
```

**Comment by ShutdownRepo** [comment]
@LOLOLEKIK Awesome, nice job!
1. In the first screenshot, we see the syntax. Imo, it should be printed without the automatic highlighting/coloring. And it would probably be best to have the syntax helper string be defined somewhere and used in the help as well as the error message, to avoid having one string different than another
2. Same thing when the ports are printed back, let's remove the color there imo
3. Maybe it would be nice to print a warning when you handle implicit syntax. For example, with `-p 3001-3005:3000` it would be relevant to raise a warning like `Ports sharing configuration could be wrong (3001-3005:3000). Assuming the following config: 3001-3005:3000-3004`
4. Also, fyi, you can specify the image name in the command line to avoid the interactive menu: `exegol start dev full -p 1234`
5. Finally, @Dramelac what do you think about supporting the last syntax? I think it would be nice to have (e.g. `5000,5001,5002,5003-5005,5010-5015`)

**Comment by LOLOLEKIK** [comment]
I've pushed some changes that display a message when the -p option is used in a non-standard way.

I've also "streamlined" the use of port sharing so that the following options give the same behaviors (see screenshot) 

```bash
-p 8000-8005:8010
-p 8010:8000-8005
```

image

Note that it's exactly for this clarity that I think it's a good idea to keep the current ouput in the container summary.

For points 1 and 2, I'm sorry but I don't think I understood everything. Alexis asked me to add the [green] flag on the output. Is there any question of removing this flag? I've tried to understand how the logger works to display raw data but I have the impression that my modifications will impact all exegol outputs.

Maybe I'm missing something obvious...

As for point 5, it's because it wasn't possible to do that that motivated me to make this PR.

(thanks for the tips :) )

File: exegol/console/cli/actions/GenericParameters.py [hunk 4]
```diff
@@ -196,7 +196,7 @@ def __init__(self, groupArgs: List[GroupArg]):
                             action="append",
                             default=[],
                             dest="ports",
-                            help="Share a network port between host and exegol (format: --port [<host_ipv4>:]<host_port>[:<container_port>][:<protocol>]. This configuration will disable the shared network with the host.",
+                            help="Share a network port between host and exegol (format: --port [<host_ipv4>:]<host_port>[-<end_host_port>][:<container_port>[-<end_container_port>]][:<protocol>]. This configuration will disable the shared network with the host.",
```
Comment (ShutdownRepo) [inline exegol/console/cli/actions/GenericParameters.py:199]: would it be possible to defined a variable called `port_sharing_syntax` and have the help dialog become `Share a network [...] (format: {port_sharing_syntax}). This configuration[...]. Then, the variable could also be used later on when raising critical messages

File: exegol/model/ContainerConfig.py [hunk 5]
```diff
@@ -1234,26 +1234,45 @@ def addUserDevice(self, user_device_config: str):
...
         If host_ipv4 is not set, default to 0.0.0.0
-        If container_port is not set, default is the same as host port
+        If container_port is not set, the same port(s) as host port(s) will be used
         If protocol is not set, default is 'tcp'"""
-        match = re.search(r"^((\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}):)?(\d+):?(\d+)?:?(udp|tcp|sctp)?$", user_test_port)
+        # Regex to capture port ranges and protocols correctly
+        match = re.search(r"^((\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}):)?(\d+)(-(\d+))?:?(\d+)?(-(\d+))?:?(udp|tcp|sctp)?$", user_test_port)
         if match is None:
-            logger.critical(f"Incorrect port syntax ({user_test_port}). Please use this format: [green][<host_ipv4>:]<host_port>[:<container_port>][:<protocol>][/green]")
+            logger.critical(f"Incorrect port syntax ({user_test_port}). Please use this format: [green][<host_ipv4>:]<host_port>[-<end_host_port>][:<container_port>[-<end_container_port>]][:<protocol>][/green]")
```
Comment (ShutdownRepo) [inline exegol/model/ContainerConfig.py:1245]: the `port_sharing_syntax` could be used here as well
The string could be placed between single quotes to have it all highlighted in green. Currently, the `[green]` flags are probably not working, the string is not all-green. This is because the string contains `[` and `]` chars.

File: exegol/model/ContainerConfig.py [hunk 6]
```diff
@@ -1234,26 +1234,45 @@ def addUserDevice(self, user_device_config: str):
...
+            end_container_port_defined = match.group(8) is not None
+            start_container_port = int(match.group(6)) if start_container_port_defined else start_host_port
+            # If start_container_port is not defined, use end_host_port, otherwise use start_container_port or end_container_port if defined
+            if not start_container_port_defined:
+                end_container_port = end_host_port
+            else:
+                end_container_port = int(match.group(8)) if match.group(8) else start_container_port
+            # check port consistency
+            if (len(range(start_host_port,end_host_port)) != len(range(start_container_port,end_container_port))) or (start_host_port != end_host_port and (not start_container_port_defined and end_container_port_defined)) or (start_host_port != end_host_port and (start_container_port_defined and not end_container_port_defined)) :
+                logger.info(f"Ports sharing configuration could be wrong ({user_test_port}). The configuration in the 'Container sumamry' below will be applied.")
```
Comment (ShutdownRepo) [inline exegol/model/ContainerConfig.py:1262]: could be wrong -> was unclear

**Review by ShutdownRepo — CHANGES_REQUESTED** [review]

File: exegol/console/cli/actions/GenericParameters.py (hunk 4 above)
Comment (LOLOLEKIK) [reply exegol/console/cli/actions/GenericParameters.py:199]: Ok, got it

File: exegol/model/ContainerConfig.py (hunk 5 above)
Comment (LOLOLEKIK) [reply exegol/model/ContainerConfig.py:1245]: Yes, i repalce the green flag by "'" only on ports

File: exegol/model/ContainerConfig.py (hunk 6 above)
Comment (LOLOLEKIK) [reply exegol/model/ContainerConfig.py:1262]: I change it

**Comment by Dramelac** [comment]
I've made some changes.
Range are now dynamically detected for greater readability.
image

> Finally, @Dramelac what do you think about supporting the last syntax? I think it would be nice to have (e.g. 5000,5001,5002,5003-5005,5010-5015)

In the current version, multiple config must be set with multiple `-p`, for example: `-p 5000-5010 -p 8089`. @ShutdownRepo Not a fan to introduce more complexity into the already long pattern...

**Review by Dramelac — APPROVED** [review]
Thank you for the addition !
I'll merge this on a dedicated dev branch for now to merge it with additional internal network rework before release.

**Review by ShutdownRepo — APPROVED** [review]
Fine by me
Didn't check everything in detail, but if something's wrong we'll see it rather quickly in dev
